addons/interactions/src/instrument.test.ts

```diff
@@ -0,0 +1,411 @@
+          },
+        },
+        options
+      );
+      expect(fn).toThrow();
+      expect(callSpy).toHaveBeenCalledWith(
+        expect.objectContaining({
+          id: '0-fn',
+          exception: {
+            name: 'Error',
+            message: 'Boom!',
+            stack: expect.stringContaining('Error: Boom!'),
+          },
+        })
+      );
+    });
+
+    it('catches thrown errors and throws an ignoredException instead', () => {
+      const { fn } = instrument(
+        {
+          fn: () => {
+            throw new Error('Boom!');
+          },
+        },
+        options
+      );
+      expect(fn).toThrow('ignoredException');
+    });
+
+    it('catches forwarded exceptions and throws an ignoredException instead', () => {
+      const { fn1, fn2 } = instrument(
+        {
+          fn1: () => {},
+          fn2: () => {
+            throw new Error('Boom!');
+          },
+        },
+        options
+      );
+      expect(() => fn1(fn2())).toThrow('ignoredException');
+    });
+  });
+
+  describe('while debugging', () => {
+    beforeEach(() => {
+      global.window.parent.__STORYBOOK_ADDON_TEST_MANAGER__.isDebugging = true;
+    });
+
+    it('defers calls to intercepted functions', () => {
+      const { fn } = instrument({ fn: jest.fn() }, { intercept: true });
+      expect(fn()).toEqual(expect.any(Promise));
+      expect(fn._original).not.toHaveBeenCalled();
+    });
+
+    it('does not defer calls to non-intercepted functions', () => {
+      const { fn } = instrument({ fn: jest.fn(() => 'ok') });
+      expect(fn()).toBe('ok');
+      expect(fn._original).toHaveBeenCalled();
+    });
+
+    it('does not defer calls to intercepted functions that are chained upon', () => {
+      const { fn1 } = instrument({ fn1: jest.fn(() => ({ fn2: jest.fn() })) }, { intercept: true });
+      global.window.parent.__STORYBOOK_ADDON_TEST_MANAGER__.chainedCallIds.add('0-fn1');
+      const res1 = fn1();
+      expect(res1.fn2()).toEqual(expect.any(Promise));
+      expect(fn1._original).toHaveBeenCalled();
+      expect(res1.fn2._original).not.toHaveBeenCalled();
+    });
+
+    it('does not defer calls while playing until a certain call', () => {
+      const { fn } = instrument({ fn: jest.fn(() => 'ok') }, { intercept: true });
+      global.window.parent.__STORYBOOK_ADDON_TEST_MANAGER__.playUntil = '1-fn';
+      /* 0-fn */ expect(fn()).toBe('ok');
+      /* 1-fn */ expect(fn()).toBe('ok');
+      /* 2-fn */ expect(fn()).toEqual(expect.any(Promise));
+      /* 3-fn */ expect(fn()).toEqual(expect.any(Promise));
+    });
+
+    it('invokes the deferred function on the "next" event', () => {
+      const { fn } = instrument({ fn: jest.fn(() => 'bar') }, { intercept: true });
+      const res = fn('foo').then((value: string) => {
+        expect(value).toBe('bar');
+        expect(fn._original).toHaveBeenCalledWith('foo');
+      });
+      addons.getChannel().emit(EVENTS.NEXT);
+      return res;
```

Comment:
```suggestion
      const res = (async () => {
        const value = await fn('foo');
        expect(value).toBe('bar');
        expect(fn._original).toHaveBeenCalledWith('foo');
      })();
      addons.getChannel().emit(EVENTS.NEXT);
      return res;
```

A suggestion to make this code look more like what the user would actually run. Another alternative:

```js
      const promise = fn('foo');
      addons.getChannel().emit(EVENTS.NEXT);
      
      const value = await promise;
      expect(value).toBe('bar');
      expect(fn._original).toHaveBeenCalledWith('foo');
```

Reply:
Yeah, I was avoiding async/await because it requires the regenerator thing for Jest and I hadn't set that up in the original repo. In the Storybook monorepo, that's already setup and so now I can actually do that.